components/ui/scroll-section-group/scroll-section-group.tsx

```diff
@@ -0,0 +1,83 @@
+import React, { useState, RefObject } from "react";
+import { cn } from "@/lib/utils";
+
+export interface ScrollSection<TSectionId extends string> {
+  id: TSectionId;
+  renderScrollSectionContent: (input: {
+    ref: RefObject<HTMLElement>,
+    key: string
+  }) => React.ReactNode;
+  ref: RefObject<HTMLElement>
+}
+
+export interface ScrollSectionGroupProps<TSectionId extends string> {
+  currentIndex: number;
+  setCurrentIndex: (input: number) => void;
+  className: string;
+  scrollThrottleThresholdMilliseconds: number;
+  sections: ScrollSection<TSectionId>[];
+  hasEnoughTimePassedSinceLastScrollEventActioned: (unixEpochTimestampMilliseconds: number) => boolean
+  setLastScrollEventActionedUnixEpochTimestampMilliseconds: (unixEpochTimestampMilliseconds: number) => void
+}
+
+enum ScrollDirection {
+  UP = 'UP',
+  DOWN = 'DOWN',
+}
+
+export const ScrollSectionGroup = <TSectionId extends string>(props: ScrollSectionGroupProps<TSectionId>): React.ReactNode => {
+  const [oldScrollTop, setOldScrollTop] = useState<number | undefined>(undefined);
+  
+  const changeViewedSectionBasedOnScroll = (currentIndex: number, scrollDirection: ScrollDirection) => {
+    const nextIndex = scrollDirection === ScrollDirection.UP ? currentIndex - 1 : currentIndex + 1;
+
+    if(nextIndex < 0 || nextIndex >= props.sections.length) {
+      return;
+    }
+
+    props.setCurrentIndex(nextIndex);
+  }
```

Comment:
Yo Sean why are we writing our own JS for scroll snapping? Isn't there a CSS solution?

Yeah there is. Unfortunately,
- From my testing the CSS solution sucks if you have large scroll snap panels and it was only working like half the time anyways. Apparently it has issues on certain browsers as well so :sparkles: _no thanks_ :sparkles:. I kept the classes like `snap-always` and `smooth-scroll` on the elements since my JS scroll can still technically break if you go crazy with the scrollbar so the CSS is more of a fallback here.
- I don't think there's a way to get both the CSS snapping and my custom route based snapping (see later comment) to play nice with each other since one would be in CSS and the other would 100% have to be in JS
- I tried some of the JS libraries for scroll snapping and I didn't like any of them so this is all homebrewed and wayyyyy better in my opinion.

I'd just try it out for yourself. Honestly, every nice looking scroll panel website I saw was either not using the CSS scroll snap at all and using pure JS or is doing what I'm doing where they are using a combination of both.


Two really nice ones I'm in love with :heart:
https://www.tesla.com/ <- CSS but clearly also using JS on top of that because I know the CSS stuff doesn't work like that
https://whereispoland.com/ <- Pure JS if you check out the DOM as you scroll

![Peek 2024-01-21 16-23](https://github.com/PathoTracker/Pathotracker/assets/86806388/70bb22cb-cc93-4338-9564-ec152ca99fe9)

Reply:
Oh yeah I also want to do some conditionals based on which page you're currently scrolled to which is a massive pain in the rear if I did this with CSS